### ISSUE TITLE
Utilize the Carbon library for localized datetime calculations

### ISSUE THREAD

**Issue by fjarrett** [affiliatewp/AffiliateWP#773]
Related to #255, #333, #399, #504, #521, and #616

#### TL;DR

[Carbon](https://github.com/briannesbitt/Carbon) makes it really easy to do complex datetime calculations in PHP with relative timezone support.

```php
/* Interval Examples */
$london = new Carbon( 'now', 'Europe/London' );
$last_week = $london->subWeek();
$two_weeks_ago = $london->subWeeks( 2 );

/* Diff Example */
$utc = new Carbon( 'now' );
$chicago = new Carbon( 'now', 'America/Chicago' );
echo $utc->diffInHours( $chicago ); // 5
```

We store everything in UTC time (which is a good thing). However, as described in #333, this creates significant accuracy problems for sites in far away timezones. Carbon allow us to unify how we calculate dates and times for use in DB queries, and how we display those dates to the end user relative to their timezone.

#### Graphs

Carbon would allow us to simplify much of our [current code](https://github.com/affiliatewp/AffiliateWP/blob/6f8557b61a15cd0d400489bf74114f7709588ee7/includes/class-graph.php#L421), that is doing a lot of homegrown calendar calculations, in favor of a cleaner object-oriented approach.

This would result in being able to extend the AffiliateWP Graph API to do pretty much any date range calculation desired, relative to the site's timezone, and possibly even allow developers to extend graphs to support their own custom date ranges or diff comparisons.

#### Other areas

Additionally, other areas of the plugin that require datetime calculations will be improved. For instance, the enhancement described in #255 for delaying commissions based on maturation time, or [this comment](https://github.com/affiliatewp/AffiliateWP/issues/504#issuecomment-70347826) in #504 where affiliates would need to be filtered by registration date and/or their total commissions within a specified date range.

**Comment by pippinsplugins** [comment]
That is so cool.

Let's seriously consider this for 1.8.

**Comment by michaelbeil** [comment]
Post 1.9, we should consider bumping this to 2.0.

**Comment by DrewAPicture** [comment]
Started a PR with a proof of concept: #2285 

Note that this work in progress is not even close to complete, simply a demonstration of how we could begin to leverage this library.

**Comment by DrewAPicture** [comment]
At this point, the PR is getting very close to feature completeness. With 40 files changed (and counting), a progress check list would be helpful. I'll add stuff as I think of what might be left:

**Under the Hood: Architecture**:
* [x] Bundle the `Carbon` library
* [x] Introduce base `\AffWP\Utils\Date` class, accessed via `utils->date`
* [ ] Full test coverage for `\AffWP\Utils\Date` methods
* [x] Introduce `utils->date()` method to retrieve a Carbon instance with the timezone already set
* [x] Create a helper to derive the timezone from `gmt_offset` if necessary
* [x] Introduce `affwp_get_filter_dates()` date filter helper as a replacement for `affwp_get_report_dates()`
* [x] Deprecate `affwp_get_report_dates()`
* [x] Introduce `affwp_get_filter_date_range()` date filter helper
* [x] Introduce `affwp_get_filter_date_values()` date filter helper for general date usage in the UI
* [ ] Full test coverage for date helpers in date-functions.php

**Under the Hood: Core UI**:
* [x] Convert the base Reports Tab class to use date filter helpers
* [x] Conver the base Graph class to use date filter helpers
* [x] Convert the Payouts graph to use date filter helpers
* [x] Convert the Registrations graph to use date filter helpers
* [x] Convert the Referrals graph to use date filter helpers
* [x] Convert the Visits graph to use date filter helpers
* [x] Introduce `affwp_date_i18n()` – a wrapper for `date_i18n()` that leverages a timestamp respecting the WP timezone

**Under the Hood: Objects**:
* [x] Adjust affiliates-db to use date helpers to convert incoming dates to UTC and outgoing to expected timezone
* [ ] Adjust creatives-db to use date helpers to convert incoming dates to UTC and outgoing to expected timezone
* [x] Adjust payouts-db to use date helpers to convert incoming dates to UTC and outgoing to expected timezone
* [x] Adjust referrals-db to use date helpers to convert incoming dates to UTC and outgoing to expected timezone
* [x] Adjust visits-db to use date helpers to convert incoming dates to UTC and outgoing to expected timezone
* [x] Introduce an `Affiliate->date_registered()` getter for either retrieving `date_registered` or a formatted version
* [x] Full test coverage for `Affiliate->date_registered()`
* [ ] Adjust creatives-db to use date helpers to convert incoming dates to UTC
* [x] Introduce a `Creative->date()` getter for either retrieving `date` or a formatted version
* [ ] Full test coverage for `Creative->date()`
* [x] Introduce a `Payout->date()` getter for either retrieving `date` or a formatted version
* [ ] Full test coverage for `Payout->date()`
* [x] Introduce a `Referral->date()` getter for either retrieving `date` or a formatted version
* [ ] Full test coverage for `Referral->date()`
* [x] Introduce a `Visit->date()` getter for either retrieving `date` or a formatted version
* [ ] Full test coverage for `Visit->date()`
* [x] Adjust `affwp_add_referral()` to use the `utils->date()` helper to set the current, formatted date

**Under the Hood: General**:
* [x] Adjust `affwp_month_num_to_name()` to use `affwp_date_i18n()`

**UI: Object Management**:
* [x] Update the Affiliate > Edit screen to output a localized/timezoned registration date
* [x] Update the Payout > Edit screen to output a localized/timezoned date
* [x] Update the Referral > Edit screen to output a localized/timezoned date
* [x] Update the Referral > New screen to output a default localized/timezoned date for 'today'
* [x] Update the "Date" column in the Payouts list table to show localized/timezoned date
* [x] Update the "Date" column in the Referrals list table to show localized/timezoned date
* [x] Update the "Date" column in the Visits list table to show localized/timezoned date
* [ ] Update the date filters in the Affiliates list table to use `affwp_get_filter_date_values()`
* [ ] Update the date filters in the Payouts list table to use `affwp_get_filter_date_values()`
* [x] Update the date filters in the Referrals list table to use `affwp_get_filter_date_values()`
* [x] Update the date filters in the Visits list table to use `affwp_get_filter_date_values()`

**UI: Tools**:
* [x] Update Payout exports to use localized/timezoned date values
* [x] Update Visit exports to use localized/timezoned date values

**UI: Affiliate Area**:
* [x] Display Payout dates as localized/timezoned
* [x] Display Referral dates as localized/timezoned
* [x] Display Visit dates as localized/timezoned

**UI: General**:
* [x] Update the license expires notice to show localized/timezoned date

**Comment by DrewAPicture** [comment]
@pippinsplugins Test coverage according to the checklist is finished and this is ready for final review!

**Comment by michaelbeil** [comment]
I am seeing a lag of two hours. For example, I just recorded a visit at 12:36, but it is 2:36.

**Comment by DrewAPicture** [comment]
@michaelbeil What timezone do you have set in your WordPress settings?

**Comment by michaelbeil** [comment]
UTC +2.
